src/thunderbots/software/test/grsim_backend/motion_controller.cpp

```diff
@@ -0,0 +1,195 @@
+        expectedAngularVelocity == roboSpeeds.second)
+    {
+        speedsEqual = true;
+    }
+    else
+    {
+        speedsEqual = false;
+    }
+
+
+    EXPECT_TRUE(speedsEqual);
+}
+
+TEST_F(MotionControllerTest, over_speed_test)
+{
+    Robot robot       = Robot(4, Point(-1, 0), Vector(2.23, 2.23), Angle::ofRadians(0),
+                        AngularVelocity::ofRadians(0), current_time);
+    double deltaTime  = 1;
+    Point destination = Point(5, 1);
+    Angle destinationAngle  = Angle::ofDegrees(0);
+    double destinationSpeed = 1.75;
+    bool speedsEqual;
+
+    std::pair<Vector, AngularVelocity> roboSpeeds = MotionController::grSimBangBang(
+        robot, destination, destinationSpeed, destinationAngle, deltaTime);
+
+    double expectedSpeed = ROBOT_MAX_SPEED;
+
+    Angle expectedAngularVelocity = Angle::ofRadians(0);
+
+    if (expectedSpeed == roboSpeeds.first.len() &&
+        expectedAngularVelocity == roboSpeeds.second)
+    {
+        speedsEqual = true;
+    }
+    else
+    {
+        speedsEqual = false;
+    }
+
+
+    EXPECT_TRUE(speedsEqual);
+}
+
+TEST_F(MotionControllerTest, no_overspeed_acceleration_test)
+{
+    Robot robot            = Robot(4, Point(-1, 0), Vector(1.3, 1.3), Angle::ofRadians(0),
+                        AngularVelocity::ofRadians(0), current_time);
+    double deltaTime       = 1;
+    Point destination      = Point(5, 1);
+    Angle destinationAngle = Angle::ofDegrees(0);
+    double destinationSpeed = 1.75;
+    bool speedsEqual;
+
+    std::pair<Vector, AngularVelocity> roboSpeeds = MotionController::grSimBangBang(
+        robot, destination, destinationSpeed, destinationAngle, deltaTime);
+
+    double expectedSpeed = ROBOT_MAX_SPEED;
+
+    Angle expectedAngularVelocity = Angle::ofRadians(0);
+
+    if (expectedSpeed == roboSpeeds.first.len() &&
+        expectedAngularVelocity == roboSpeeds.second)
+    {
+        speedsEqual = true;
+    }
+    else
+    {
+        speedsEqual = false;
+    }
+
+
+    EXPECT_TRUE(speedsEqual);
+}
+
+TEST_F(MotionControllerTest, no_overspeed_ang_acceleration_test)
+{
+    Robot robot             = Robot(4, Point(-1, -1), Vector(0, 0), Angle::ofRadians(0),
+                        AngularVelocity::ofRadians(3.9), current_time);
+    double deltaTime        = 1;
+    Point destination       = Point(-1, -1);
+    Angle destinationAngle  = Angle::ofDegrees(210);
+    double destinationSpeed = 0;
+    bool speedsEqual;
+
+    std::pair<Vector, AngularVelocity> roboSpeeds = MotionController::grSimBangBang(
+        robot, destination, destinationSpeed, destinationAngle, deltaTime);
+
+    double expectedSpeed = 0;
+
+    Angle expectedAngularVelocity = Angle::ofRadians(4);
+
+    if (expectedSpeed == roboSpeeds.first.len() &&
+        expectedAngularVelocity == roboSpeeds.second)
+    {
+        speedsEqual = true;
+    }
+    else
+    {
+        speedsEqual = false;
+    }
+
+
+    EXPECT_TRUE(speedsEqual);
+}
+
```

Comment:
We should probably add some more test cases here. A good start would be combinations of
- Robot ahead of dest in x and y, robot behind dest in x and y, robot ahead of x and behind y, robot behind x and in front of x and y
- Negative desired final speed, zero final speed, positive desired final speed
- Negative desired final orientation, zero final speed, positive desired final orientation
- Negative delta_time, 0 delta time, positive delta_time

There are obviously a bunch of things here, but since the tests are basically all the same here, we can use parameterized tests to do this really efficiently. Parameterized tests basically define the test and then let you just change the inputs to it.  Here's the gtest documentation, please let me know if you want any help at all with this, I'd be happy to help! https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#value-parameterized-tests

Reply:
Will do

Reply:
Added 25 new unit tests.